### PR TITLE
StripeUICore: changed OneTimeTextField to be more configurable and removed its dot UI

### PR DESCRIPTION
**Pull request by kgaidis-stripe**
## Summary

This PR makes changes to the OTP field (`OneTimeCodeTextField`):
1. It makes it more configurable
2. It removes the "middle dot"
3. It changes caret color to be the color of the text (...because that logically makes sense and is that way in designs)
4. It changes the sizing of the caret to be slightly smaller

Many of these changes should be a no-op as far as the previous state goes.

[Link to designs I referenced](https://www.figma.com/file/wXQ8NJApqSJiLmxxANDRTs/%E2%9C%A8-Connections-3.0?type=design&node-id=2494-170914&mode=design&t=EeVqFLGzz77UhO7B-0)

### FC Testing

| Before | After |
| --- | --- |
| before_fc | after_fc |

#### Before Video

https://github.com/stripe/stripe-ios/assets/105514761/2ed8aa15-6f19-434d-8ae1-5d7a65be19f0

#### After Video

https://github.com/stripe/stripe-ios/assets/105514761/8ebd5d50-1eb0-4f2d-864e-c10f89803f65

I ran Financial Connections end-to-end tests which utilize the OTP field:

```
bitrise run financial-connections-stability-tests

FinancialConnectionsNetworkingUITests
    ✓ testNativeNetworkingTestMode (80.792 seconds)
FinancialConnectionsUITests
    ✓ testDataLiveModeOAuthNativeAuthFlow (34.097 seconds)
    ✓ testDataLiveModeOAuthWebAuthFlow (26.395 seconds)
    ✓ testDataTestModeOAuthNativeAuthFlow (23.219 seconds)
    ✓ testPaymentTestModeLegacyNativeAuthFlow (25.056 seconds)
    ✓ testPaymentTestModeManualEntryNativeAuthFlow (28.331 seconds)
    ✓ testSearchInLiveModeNativeAuthFlow (26.679 seconds)
```

This is a screenshot where I modified the field in a custom testing branch (to test customizations):

```
private(set) lazy var otpTextField: OneTimeCodeTextField = {
        let otpTextField = OneTimeCodeTextField(theme: theme)
        let otpTextField = OneTimeCodeTextField(
            configuration: OneTimeCodeTextField.Configuration(
                itemSpacing: 8,
                enableDigitGrouping: false,
                font: UIFont.systemFont(ofSize: 28, weight: .regular),
                itemCornerRadius: 12,
                itemHeight: 56
            ),
            theme: theme
        )
```

OTP_modifications

### Identity Testing

| Before | After |
| --- | --- |
| Simulator Screenshot - iPhone 12 mini - 2024-01-23 at 10 50 05 | after_identity |

### CODE REVIEW

File: StripeUICore/StripeUICore/Source/Controls/OneTimeCodeTextField.swift [hunk 1]
```diff
@@ -88,9 +133,12 @@ import UIKit
     public lazy var tokenizer: UITextInputTokenizer = UITextInputStringTokenizer(textInput: self)
 
     // MARK: -
-    public init(numberOfDigits: Int = 6, theme: ElementsUITheme = .default) {
-        self.numberOfDigits = numberOfDigits
-        self.textStorage = TextStorage(capacity: numberOfDigits)
+    public init(
+        configuration: Configuration = Configuration(),
+        theme: ElementsUITheme = .default
```
Comment (kgaidis-stripe) [inline StripeUICore/StripeUICore/Source/Controls/OneTimeCodeTextField.swift:138]: it likely makes sense to remove `theme` and put it into `configuration` but taking it one-step-at-a-time and this current PR should get the job done for now

File: StripeUICore/StripeUICore/Source/Controls/OneTimeCodeTextField.swift [hunk 2]
```diff
@@ -648,45 +702,37 @@ private extension OneTimeCodeTextField {
...
         private lazy var focusRing: CALayer = {
             let focusRing = CALayer()
-            focusRing.borderWidth = Constants.focusRingThickness
-            focusRing.cornerRadius = Constants.cornerRadius + (Constants.focusRingThickness / 2)
+            focusRing.borderWidth = configuration.focusRingThickness
+            focusRing.cornerRadius = configuration.cornerRadius + (configuration.focusRingThickness / 2)
             return focusRing
         }()
 
-        private lazy var dot: CALayer = {
```
Comment (kgaidis-stripe) [inline StripeUICore/StripeUICore/Source/Controls/OneTimeCodeTextField.swift:669]: this is the code deletion that removes the dots

File: StripeUICore/StripeUICore/Source/Controls/OneTimeCodeTextField.swift [hunk 3]
```diff
@@ -648,45 +702,37 @@ private extension OneTimeCodeTextField {
...
             label.isAccessibilityElement = false
             label.textColor = theme.colors.textFieldText
-            label.font = font
+            label.font = configuration.font
             return label
         }()
 
         var caretRect: CGRect {
-            let caretSize = CGSize(width: 2, height: font.ascender - font.descender)
+            let caretSize = CGSize(width: 2, height: configuration.font.ascender)
```
Comment (kgaidis-stripe) [inline StripeUICore/StripeUICore/Source/Controls/OneTimeCodeTextField.swift:735]: this is the change that modifies the caret size to be a little shorter

File: StripeUICore/StripeUICore/Source/Controls/OneTimeCodeTextField.swift [hunk 4]
```diff
@@ -761,8 +801,7 @@ private extension OneTimeCodeTextField {
         private func updateColors() {
             borderLayer.backgroundColor = isEnabled ? theme.colors.background.cgColor : theme.colors.disabledBackground.cgColor
             borderLayer.borderColor = theme.colors.border.cgColor
-            dot.backgroundColor = Constants.dotColor.cgColor
-            caret.backgroundColor = tintColor.cgColor
+            caret.backgroundColor = label.textColor.cgColor
```
Comment (kgaidis-stripe) [inline StripeUICore/StripeUICore/Source/Controls/OneTimeCodeTextField.swift:804]: this changes the color of the caret

**Review by ccen-stripe — APPROVED** [review]

**Pull request merged**